File: linked_lists/single_linked_lists.py
class Node:
    def __init__(self, data: int, next=None):
        self.data = data
        self.next = next
    
    def __str__(self) -> str:
        return f"data: {self.data}, next: {self.next}"


class LinkedList: 
    def __init__(self):
        self.head = None


    def append(self, new_node: Node):
        new_node.next = self.head
        self.head = new_node


    def push(self, new_node: Node):
        if self.head is None:
            self.head = new_node
            return
        last = self.head
        while(last.next):
            last = last.next
        last.next = new_node


    def insert(self, new_node: Node, index: int) -> None:
        if index == 0:
            new_node.next = self.head
            self.head = new_node
            return
        current = self.head
        for _ in range(index - 1):
            if current.next:
                current = current.next
        if current:
            new_node.next = current.next
            current.next = new_node


    def delete(self, index: int) -> int|None:
        deleted = None
        previous = None
        if index == 0:
            deleted = self.head.data
            self.head = self.head.next
            return deleted

        current = self.head
        for _ in range(index):
            if current:
                previous = current
                current = current.next

        if current:
            deleted = current.data
            previous.next = current.next

        return deleted


    def reverse(self):
        prev = None
        current = self.head
        while(current is not None):
            # next = current.next # 2
            # current.next = prev # None
            # prev = current # 1
            # current = next # 2
            current.next, prev, current = prev, current, current.next
        self.head = prev


if __name__ == "__main__":
    l = LinkedList()
    
    l.push(Node(1))
    l.append(Node(2))
    l.insert(Node(3), 1)
    print(l.head)

    l.reverse()
    print(l.head)

    print(l.delete(1))
    print(l.head)

File: linked_lists/single_linked_lists.go
package main

import "fmt"

type Node struct {
	data int
	next *Node
}

type LinkedList struct {
	head *Node
}

func (l *LinkedList) Append(newNode Node) {
	newNode.next = l.head
	l.head = &newNode
}

func (l *LinkedList) Push(newNode Node) {
	if l.head == nil {
		l.head = &newNode
		return
	}
	last := l.head
	for last.next != nil {
		last = last.next
	}
	last.next = &newNode
}

func (l *LinkedList) Insert(newNode Node, index int) {
	if index == 0 {
		newNode.next = l.head
		l.head = &newNode
		return
	}
	current := l.head
	for i := 0; current.next != nil && i < index; i++ {
		current = current.next
	}
	if current != nil {
		newNode.next = current.next
		current.next = &newNode
	}
}

func (l *LinkedList) Delete(index int) int {
	var deleted int;
	if index == 0 {
		deleted = l.head.data
		l.head = l.head.next
		fmt.Println(deleted)
		return deleted
	}

	var previous *Node
	current := l.head
	for i := 0; current.next != nil && i < index; i++ {
		previous = current
		current = current.next
	}
	if current != nil {
		deleted = current.next.data
		previous.next = current.next
	}

	return deleted
}

func (l *LinkedList) Reverse() {
	var previous *Node
	current := l.head
	for current != nil {
		next := current.next
		current.next = previous
		previous = current
		current = next
	}
	l.head = previous
}

func (l *LinkedList) Print() {
	node := l.head
	for node != nil {
		fmt.Printf("data: %d, next: %o | ", node.data, node.next)
		node = node.next
	}
	fmt.Println("\n")
}

func main() {
	var l LinkedList
	l.Push(Node{data: 1})
	l.Append(Node{data: 2})
	l.Insert(Node{data: 3}, 1)
	l.Print()

	l.Reverse()
	l.Print()

	del := l.Delete(1)
	fmt.Println(del)
	l.Print()
}